server/rest/src/main/java/org/infinispan/rest/server/operations/CacheOperations.java

```diff
@@ -0,0 +1,303 @@
+import org.infinispan.rest.MimeMetadata;
+import org.infinispan.rest.MimeMetadataBuilder;
+import org.infinispan.rest.RestCacheManager;
+import org.infinispan.rest.configuration.RestServerConfiguration;
+import org.infinispan.rest.server.CacheControl;
+import org.infinispan.rest.server.InfinispanRequest;
+import org.infinispan.rest.server.InfinispanResponse;
+import org.infinispan.rest.server.operations.exceptions.NoCacheFoundException;
+import org.infinispan.rest.server.operations.exceptions.NoDataFoundException;
+import org.infinispan.rest.server.operations.exceptions.NoKeyException;
+import org.infinispan.rest.server.operations.exceptions.NoMediaTypeException;
+import org.infinispan.rest.server.operations.exceptions.RestResponseException;
+import org.infinispan.rest.server.operations.mediatypes.MediaType;
+
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+public class CacheOperations {
+
+   private static final MurmurHash3 hashFunc = MurmurHash3.getInstance();
+
+   private final RestCacheManager restCacheManager;
+   private final RestServerConfiguration restServerConfiguration;
+
+   public CacheOperations(RestServerConfiguration configuration, RestCacheManager cacheManager) {
+      this.restServerConfiguration = configuration;
+      this.restCacheManager = cacheManager;
+   }
+
+   public InfinispanResponse getCacheValues(InfinispanRequest request) throws RestResponseException {
+      try {
+         String cacheName = request.getCacheName();
+         AdvancedCache<String, byte[]> cache = restCacheManager.getCache(cacheName);
+         CacheSet<String> keys = cache.keySet();
+         MediaType mediaType = request.getMediaType()
+               .map(m -> MediaType.fromMediaTypeAsString(m))
+               .orElseThrow(NoMediaTypeException::new);
+
+         InfinispanResponse response = InfinispanResponse.inReplyTo(request);
+         response.contentAsBytes(mediaType.getOutputPrinter().print(cacheName, keys));
+         return response;
+      } catch (CacheException cacheException) {
+         throw new NoCacheFoundException(cacheException.getLocalizedMessage());
+      }
+   }
+
+   public InfinispanResponse getCacheValue(InfinispanRequest request) throws RestResponseException {
+      try {
+         String cacheName = request.getCacheName();
+         String key = request.getKey().orElseThrow(NoKeyException::new);
+         String cacheControl = request.getCacheControl().orElse("");
+         boolean returnBody = request.getRawRequest().method() == HttpMethod.GET;
+         CacheEntry<String, byte[]> entry = restCacheManager.getInternalEntry(cacheName, key);
+         MediaType mediaType = request.getMediaType()
+               .map(m -> MediaType.fromMediaTypeAsString(m))
+               .orElse(MediaType.TEXT_PLAIN);
+
+         InfinispanResponse response = InfinispanResponse.inReplyTo(request);
+         response.status(HttpResponseStatus.NOT_FOUND);
```

Comment:
Wonder if the Response should just start as `NOT_FOUND` in InfinispanResponse. That or we can just put it in the 2 branches below where it would be used.